cirq/value/tensor_product_state.py

```diff
@@ -0,0 +1,345 @@
+
+    def __mul__(self, other):
+        if not isinstance(other, TensorProductState):
+            raise ValueError("Multiplication is only supported "
+                             "with other TensorProductStates.")
+
+        dupe_qubits = set(other.states.keys()) & set(self.states.keys())
+        if len(dupe_qubits) != 0:
+            raise ValueError(
+                "You tried to tensor two states, "
+                "but both contain factors for these qubits: {}".format(
+                    sorted(dupe_qubits)))
+
+        new_states = self.states
+        new_states.update(other.states)
+        return TensorProductState(new_states)
+
+    def __str__(self):
+        return ' * '.join(f'{st}({q})' for q, st in self.states.items())
+
+    def __repr__(self):
+        states_dict_repr = ', '.join(
+            f'{repr(key)}: {repr(val)}' for key, val in self.states.items())
+        return 'cirq.TensorProductState({%s})' % states_dict_repr
+
+    def __getitem__(self, qubit):
+        """Return the NamedState at the given qubit."""
+        return self.states[qubit]
+
+    def __iter__(self):
+        yield from self.states.items()
+
+    def __len__(self):
+        return len(self.states)
+
+    def __eq__(self, other):
+        if not isinstance(other, TensorProductState):
+            return False
+
+        return self.states == other.states
+
+    def __hash__(self):
+        return hash(tuple(self.states.items()))
+
+    def _json_dict_(self):
+        return {
+            'cirq_type': self.__class__.__name__,
+            'states': list(self.states.items()),
+        }
+
+    @classmethod
+    def _from_json_dict_(cls, states, **kwargs):
+        return cls(states=dict(states))
+
+    def state_vector(self, qubit_order: 'cirq.QubitOrder' = None):
+        """The state-vector representation of this state."""
+        from cirq import ops
+        if qubit_order is None:
+            qubit_order = ops.QubitOrder.DEFAULT
+        qubit_order = ops.QubitOrder.as_qubit_order(qubit_order)
+        qubits = qubit_order.order_for(self.qubits)
+
+        mat = 1.0 + 0.j
+        for qubit in qubits:
+            oneq_state = self[qubit]
+            state_vector = oneq_state.state_vector()
+            mat = np.kron(mat, state_vector)
+
+        return mat
+
+    def projector(self, qubit_order: 'cirq.QubitOrder' = None):
+        """The projector associated with this state expressed as a matrix.
+
+        This is |s><s| where |s> is this state.
+        """
+        from cirq import ops
+        if qubit_order is None:
+            qubit_order = ops.QubitOrder.DEFAULT
+        qubit_order = ops.QubitOrder.as_qubit_order(qubit_order)
+        qubits = qubit_order.order_for(self.qubits)
+
+        mat = 1.0 + 0.j
+        for qubit in qubits:
+            oneq_state = self[qubit]
+            oneq_proj = oneq_state.projector()
+            mat = np.kron(mat, oneq_proj)
+        return mat
+
+
+class _KetPlus(_Named1QState):
```

Comment:
What is the benefit of having `_Named1QState` as an abstract base class and then defining each of the derived classes individually over, say, the `_Named1QState` base class taking in 4 arguments (str, repr, state_vector, stabilized_by) and then directly defining `KET_PLUS` etc. as objects of the `_Named1QState` base class? The later has an advantage of shorter code and reduced boilerplate. 

IIUC, the derived classes `_KetPlus` etc. are not really singletons here (`id(cirq.KET_PLUS) != id(cirq.value.tensor_product_state._KetPlus())`) but the use of `__class__` in the base class makes it work fine. Since we are anyways exposing only the final initialized variables, I'm curious to understand the benefits of this approach versus differentiating the different objects simply based on parameters passed to the base class, as described above.

Reply:
NVM, found the answer here: https://github.com/quantumlib/Cirq/pull/2781#discussion_r459166296

> big problem: stabilized by needs to import cirq.ops which would cause a circular import because of value_equality.

This could be addressed by passing an index to `_Named1QState` and using `ops.Pauli.by_index()`

> Small problem: we'd be initializing the arrays at import time. We don't do this anywhere else

Can you provide link to relevant reading material which explains the difference? (tried searching for relevant keywords but couldn't find anything relevant).

Reply:
Thanks for linking to the previous thread. I'll think about this suggestion very seriously since you and @Strilanc both raised it :) 

Like I said, this whole file mimics the structure of the Pauli gates (i.e. `cirq.X`)

> IIUC, the derived classes _KetPlus etc. are not really singletons here
Sure; this is a Python limitation as far as I can tell. They're idiomatic singletons

> This could be addressed by passing an index to _Named1QState and using ops.Pauli.by_index()

This is true and something I will seriously consider. I think then I'd have to actually refactor `_Named1QState` into an abstract base class `OneQubitState` and the suggested class `PauliOneQubitState`

> Can you provide link to relevant reading material which explains the difference?

I suppose you could briefly read how Python's import system works. In summary: when you import a "module" the file is executed in its entirety. Because most files are just a bunch of function `def`s, you don't see this in action. But! this behavior is why you need to put `if __name__ == '__main__'` in scripts; otherwise every time you import that module it would execute the script! If you have a global constant e.g. `MY_ARRAY = np.zeros(50000)` it would actually allocate the array when you `import cirq` instead of how gates (and current states) work when an array is only allocated when you call `unitary()` or `state_vector()`. Obviously these arrays for states are tiny, so it's probably not a huge deal in practice but I was just noting that we don't do this elsewhere. For example, the X gate is *not* defined as

```python
X = MatrixGate(np.array([[0,1],[1,0]]))
```

Reply:
here's a strawman implementation
```python
# @value_equality
class PauliOneQubitState(OneQubitState):
    def __init__(self, pauli:'cirq.Pauli', eigenvalue:int):
        if eigenvalue not in [-1, 1]:
            raise ValueError("Eigenvalue must be -1 or +1")

        self.pauli = pauli
        self.eigenvalue=eigenvalue

    def _value_equality_values_(self):
        return (self.pauli, self.eigenvalue)

    def _json_dict_(self):
        return protocols.obj_to_dict_helper(self, ['pauli', 'eigenvalue'])

    def __str__(self):
        pm = '+' if self.eigenvalue == 1 else '-'
        return f'{pm}{self.pauli}'

    def state_vector(self):
        from cirq.ops.pauli_gates import _PauliX, _PauliZ
        if isinstance(self.pauli, _PauliX):
            if self.eigenvalue == 1:
                return np.array([1, 1]) / np.sqrt(2)
            elif self.eigenvalue == -1:
                return np.array([1, -1]) / np.sqrt(2)
            else:
                raise ValueError("Eigenvalue must be -1 or +1")
        if isinstance(self.pauli, _PauliZ):
            if self.eigenvalue == 1:
                return np.array([1,0])
            elif self.eigenvalue == -1:
                return np.array([0,1])
            else:
                raise ValueError("Eigenvalue must be -1 or +1")
        # ...
```

at the end of the day we need to explicitly write out the state vectors for these states just like how we have to write the unitaries for named gates somewhere, so I'm still leaning towards modeling this after how `cirq.X` is made

Reply:
^ Thanks for the explanation! For some reason, I completely overlooked the `state_vector` arrays in argument and was thinking about `PAULI_STATES` list below. My bad.

Regarding the implementation above, if we have strong opinions about not initialising the `state_vector` array at import time, I would also prefer the original implementation similar to `cirq.X` instead of the above one where we check all cases in `state_vector` function. 

My original suggestion was mainly to reduce code redundancy. To get around the problem of allocating numpy arrays, we could also wrap the array in a lambda while passing it as an argument to the class.  eg: 

```python
KET_PLUS = _Named1QState(
    name='KET_PLUS',
    str='+X',
    state_vector = lambda: np.array([1,1])/np.sqrt(2),  # state_vector wrapped in a lambda to avoid allocation at import time.
    stabilized_by=(+1, 0) # 0 instead of ops.X to avoid circular import
)
```


Again, I don't have very strong opinions here. Maybe @Strilanc can chime in since he also originally pointed this out.

Reply:
yeah your solution looks nice as well. I don't have strong preferences. Consensus should be reached on these questions:

1. Do we actually care that the current implementation is fairly verbose and produces singletons that aren't so single?
2. Do we actually care to avoid making numpy arrays at import time?
3. Do we mind that passing a pauli operator index instead of using `ops.X` is mildly less readable?

Maybe we can discuss briefly at cirq cync if people don't comment here

Reply:
Discussed over video call: going to stick with current implementation but make the base class public so others can inherit from it. (side note: maybe we should make the pauli base class public (for gates) #1985)